py/desispec/coaddition.py

```diff
@@ -702,62 +737,37 @@ def coadd_cameras(spectra, cosmics_nsig=0., onetile=False) :
-                    ttflux[1:] = ttflux[1:]-ttflux[:-1]
-                    ttvar[1:]  = ttvar[1:]+ttvar[:-1]
-                    ttflux[0]  = 0
-                    grad.append(ttflux)
-                    gradvar.append(ttvar)
-
             ivar_unmasked[i,windices] += np.sum(spectra.ivar[b][jj],axis=0)
 
             if spectra.mask is not None :
                 ivarjj=spectra.ivar[b][jj]*(spectra.mask[b][jj]==0)
             else :
                 ivarjj=spectra.ivar[b][jj]
-
             if cosmics_nsig is not None and cosmics_nsig > 0 and len(jj)>2  :
-                grad=np.array(grad)
-                gradivar=1/np.array(gradvar)
-                nspec=grad.shape[0]
-                meangrad=np.sum(gradivar*grad,axis=0)/np.sum(gradivar)
-                deltagrad=grad-meangrad
-                chi2=np.sum(gradivar*deltagrad**2,axis=0)/(nspec-1)
-                bad  = (chi2>cosmics_nsig**2)
-                nbad = np.sum(bad)
-                if nbad>0 :
-                    log.info("masking {} values for targetid={}".format(nbad,tid))
-                    badindex=np.where(bad)[0]
-                    for bi in badindex  :
-                        k=np.argmax(gradivar[:,bi]*deltagrad[:,bi]**2)
-                        ivarjj[k,bi]=0.
-                        log.debug("masking spec {} wave={}".format(k,spectra.wave[b][bi]))
+                _mask_cosmics(spectra.wave[b], spectra.flux[b],
+                                              spectra.ivar[b],
+                                              spectra.mask[b] if spectra.mask is not None else None,
+                                              jj, ivarjj, cosmics_nsig=cosmics_nsig,
+                              tid=tid)
 
             ivar[i,windices] += np.sum(ivarjj,axis=0)
             flux[i,windices] += np.sum(ivarjj*spectra.flux[b][jj],axis=0)
+
             if spectra.resolution_data is not None:
-                for r in range(band_ndiag) :
-                    rdata[i,r+(ndiag-band_ndiag)//2,windices] += np.sum((spectra.ivar[b][jj]*spectra.resolution_data[b][jj,r]),axis=0)
+                ww = spectra.resolution_data[b].shape[1]//2
+                # resolution kernel width
+                npix = spectra.resolution_data[b].shape[2]
+                # indices of the corresponding variance point
+                # that needs to be used for ivar weights
+                res_indices = (np.arange(npix)[None, :]
+                               + np.arange(-ww, ww+1)[:, None]) % npix
+                res_whts = np.array([_[res_indices] for _ in ivarjj])
+                for r in range(band_ndiag):
+                    cur_off =  (ndiag - band_ndiag)//2
+                    rdata[i, r + cur_off, windices] += np.sum(res_whts[:, r] *
+                                        spectra.resolution_data[b][jj, r , : ],
+                                                   axis=0)
+                    rdata_norm[i, r + cur_off, windices] += res_whts.sum(axis=0)[r]
```

Comment:
This implementation in `coadd_cameras` is very similar to the implementation in `coadd`, but differs in details: ivarjj_masked vs. ivarjj and exactly how the resolution data (trdata, rdata, rdata_norm...) are updated.  Could you combine these into a single function that is called by both `coadd` and `coadd_cameras` to guarantee that they are doing the same thing?  i.e. similar to what you did for `_mask_cosmics`.

Reply:
Yes, I'll try to do that.
I also was trying to update the code (already done with coadd())  how it deals with the mask=None, vs not to avoid having two separate code-paths in multiple places
 there (which was the cause of some issues for coadd_cameras)